### ISSUE TITLE
overflow during drop-check on a recursive type

### ISSUE THREAD

**Issue by darkstalker** [rust-lang/rust#44933]
I get a pretty strange error when trying to compile this: ([playground link](https://play.rust-lang.org/?gist=5490697a4e08dd949d7a6abbc58a2bd1&version=stable))
```Rust
enum Stuff<T> {
    Value(T),
    Nested(Box<Stuff<Stuff<T>>>),
}

fn main() {
    let a = Stuff::Value(1);
}
```
output:
```
error[E0320]: overflow while adding drop-check rules for Stuff<i32>
 --> src/main.rs:7:9
  |
7 |     let a = Stuff::Value(1);
  |         ^
  |
  = note: overflowed on Stuff<Stuff<Stuff<Stuff<Stuff<Stuff< ...
```
If the type isn't used, the program compiles fine. Don't know if the type is actually representable or not, the error is misleading.

**Comment by sfackler** [comment]
How is it misleading? `Stuff::<i32>::Nested` contains a `Stuff<Stuff<i32>>`, which contains a `Stuff<Stuff<Stuff<i32>>>`, which contains a....

**Comment by Ixrec** [comment]
I wouldn't call it _misleading_ either, but "overflow while adding drop-check rules" certainly isn't as clear and to the point as, say, "Stuff<T> is unrepresentable". It also seems bad that we don't emit this error until the type gets used somewhere.

**Comment by pnkfelix** [comment]
(I am debating about whether the I-hang label applies here. In my head, stack overflow errors are usually just the result of hangs being caught via very crude means...)

**Comment by pnkfelix** [comment]
Interestingly, the overflow here doesn't trigger an ICE?

**Comment by oli-obk** [comment]
Diagnostics triage: P-low, as it's fairly clear what's going on, just not a great diagnostic message

**Comment by estebank** [comment]
Triage: no change.

**Comment by Dylan-DPC** [comment]
Current error:
```
error[E0320]: overflow while adding drop-check rules for Stuff<i32>
 --> src/main.rs:9:9
  |
9 |     let a = Stuff::Value(1);
  |         ^
  |
  = note: overflowed on Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<Stuff<i32>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

For more information about this error, try `rustc --explain E0320`.
```

**Comment by estebank** [comment]
@Dylan-DPC surprised that we don't trim the type name in the `note` and store it to disk. That would be an easy improvement, as well as modifying the message to surround the type name with `.

**Comment by ejjonny** [comment]
I've run into a similar error with a type that looks to me to be completely reasonable & only fails to compile in some cases. Perhaps the triage for this issue can be revisited.

```rust
enum Tree<T: Scopable> {
    Group(Vec<Tree<T>>),
    Subtree(Box<Tree<T::SubType>>),
    Leaf(T),
}

trait Scopable: Sized {
    type SubType: Scopable;
}

impl<T: Scopable> Tree<T> {
    fn foo(self) -> Self { // error[E0320]: overflow while adding drop-check rules for Tree<T>
        self
    }
}
```

I was hoping to implement a builder API on `Tree`, but I don't know any way around this error.

I did notice that using references in `foo` do not trigger the drop check overflow, but I can't implement a builder pattern without owned values.
